File: boilerplate/java/order-service/src/main/java/com/example/orderservice/application/usecases/RefreshTokenUseCaseImpl.java
package com.example.orderservice.application.usecases;

import com.example.orderservice.application.dtos.RefreshTokenCommand;
import com.example.orderservice.application.dtos.RefreshTokenResult;
import com.example.orderservice.domain.models.AuthenticationException;
import com.example.orderservice.domain.models.User;
import com.example.orderservice.domain.models.UserId;
import com.example.orderservice.domain.ports.TokenGenerator;
import com.example.orderservice.domain.ports.TokenParser;
import com.example.orderservice.domain.ports.UserRepository;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class RefreshTokenUseCaseImpl implements RefreshTokenUseCase {

    private final TokenParser tokenParser;
    private final TokenGenerator tokenGenerator;
    private final UserRepository userRepository;

    public RefreshTokenUseCaseImpl(
            TokenParser tokenParser,
            TokenGenerator tokenGenerator,
            UserRepository userRepository) {
        this.tokenParser = tokenParser;
        this.tokenGenerator = tokenGenerator;
        this.userRepository = userRepository;
    }

    @Override
    public RefreshTokenResult execute(RefreshTokenCommand command) {
        if (command == null || command.refreshToken() == null || command.refreshToken().isBlank()) {
            throw new IllegalArgumentException("refreshToken must not be null or blank");
        }

        // Validate the refresh token and extract user ID
        if (!tokenParser.isValid(command.refreshToken())) {
            throw new AuthenticationException("AUTH_INVALID_TOKEN", "Invalid or expired refresh token");
        }

        Optional<UserId> userIdOpt = tokenParser.parseUserId(command.refreshToken());
        if (userIdOpt.isEmpty()) {
            throw new AuthenticationException("AUTH_INVALID_TOKEN", "Could not parse user from token");
        }

        UserId userId = userIdOpt.get();
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new AuthenticationException("AUTH_USER_NOT_FOUND", "User not found"));

        // Generate new token pair (token rotation)
        String newAccessToken = tokenGenerator.generateAccessToken(user);
        String newRefreshToken = tokenGenerator.generateRefreshToken(user);

        Set<String> roles = user.getRoles().stream()
            .map(r -> r.getCode())
            .collect(Collectors.toSet());

        return new RefreshTokenResult(
            newAccessToken,
            newRefreshToken,
            user.getEmail().getValue(),
            roles,
            "Bearer"
        );
    }
}

File: boilerplate/nestjs/order-service/src/application/usecases/refresh-token.use-case.impl.ts
import { Injectable, Inject } from '@nestjs/common';

import { IRefreshTokenUseCase } from './refresh-token.use-case.interface';
import { AuthenticationException } from '../../domain/exceptions/auth.exception';
import { ITokenGenerator } from '../../domain/ports/token-generator.port';
import { ITokenParser } from '../../domain/ports/token-parser.port';
import { IUserRepository } from '../../domain/ports/user-repository.port';
import { RefreshTokenCommand, RefreshTokenResult } from '../dtos/refresh-token.dto';

@Injectable()
export class RefreshTokenUseCaseImpl implements IRefreshTokenUseCase {
    constructor(
        @Inject('ITokenParser')
        private readonly tokenParser: ITokenParser,
        @Inject('ITokenGenerator')
        private readonly tokenGenerator: ITokenGenerator,
        @Inject('IUserRepository')
        private readonly userRepository: IUserRepository,
    ) {}

    async execute(command: RefreshTokenCommand): Promise<RefreshTokenResult> {
        if (!command?.refreshToken) {
            throw new Error('refreshToken must not be null or blank');
        }

        if (!this.tokenParser.isValid(command.refreshToken)) {
            throw new AuthenticationException('AUTH_INVALID_TOKEN', 'Invalid or expired refresh token');
        }

        const userId = this.tokenParser.parseUserId(command.refreshToken);
        if (!userId) {
            throw new AuthenticationException('AUTH_INVALID_TOKEN', 'Could not parse user from token');
        }

        const user = await this.userRepository.findById(userId);
        if (!user) {
            throw new AuthenticationException('AUTH_USER_NOT_FOUND', 'User not found');
        }

        // Token rotation: generate new pair
        const newAccessToken = this.tokenGenerator.generateAccessToken(user);
        const newRefreshToken = this.tokenGenerator.generateRefreshToken(user);

        return {
            accessToken: newAccessToken,
            refreshToken: newRefreshToken,
            email: user.email.getValue(),
            roles: Array.from(user.roles),
            tokenType: 'Bearer',
        };
    }
}
